src/api_v2/api.rs

```diff
@@ -0,0 +1,143 @@
+    }
+
+    ///
+    /// Allows reading of the public profile of the user (/me).
+    ///
+    /// `identify`  is the default scope for the Authorization Code Grant and always implicitly provided.
+    pub fn identify(self) -> Self {
+        return self.add_scope("identify");
+    }
+
+    /// Allows reading of publicly available data on behalf of the user.
+    pub fn public(self) -> Self {
+        return self.add_scope("public");
+    }
+
+    fn add_scope(mut self, scope: &str) -> Self {
+        let s = &self.str;
+        self.str = format!("{s} {scope}");
+        return self;
+    }
+
+    fn to_string(self) -> String {
+        return self.str.to_string();
+    }
+}
+
+const API_URL: &str = "https://osu.ppy.sh/oauth/";
+const API_AUTHORIZE: &str = "authorize";
+const API_TOKEN: &str = "token";
+
+pub struct Api {
+    bot: Arc<User>,
+    client: reqwest::Client,
+    redirect_uri: String,
+}
+
+impl Api {
+    /// url is application callback url
+    /// make sure callback url right
+    pub fn new(client_id: i32, code: String, url: String) -> Result<Self, Error> {
+        let bot = User::create_bot(code, client_id);
+        let bot = Arc::new(bot);
+        let client = reqwest::Client::new();
+        let redirect_uri = url.to_string();
+        Ok(Self {
+            bot,
+            client,
+            redirect_uri,
+        })
+    }
+
+    pub fn get_oauth_url(&self, scopes: Scopes, state: &str) -> String {
+        let client_id = &self.bot.uid;
+        let redirect_uri = &self.redirect_uri;
+        let response_type = "code";
+        let scope = scopes.to_string();
+        let url = format!(
+            "{API_URL}{API_AUTHORIZE}?\
+            client_id={client_id}&\
+            redirect_uri={redirect_uri}&\
+            response_type={response_type}&\
+            scope={scope}&\
+            state={state}");
+        return url;
+    }
+
+    pub async fn refresh_token(&self, user: &mut User) {
+        let url = format!("{API_URL}{API_TOKEN}");
+        let header = self.header(user);
+        let body = json!({
+            "client_id": self.bot.uid.to_string(),
+            "client_secret": self.bot.refresh_token.to_string(),
+            "refresh_token": user.refresh_token.to_string(),
+            "grant_type": "refresh_token".to_string(),
+            "redirect_uri": self.redirect_uri.clone()
+        });
+
+        let mut request_build = self.client.post(url)
+            .body(body.to_string());
+        for (k, v) in header {
+            request_build = request_build.header(k, v);
+        }
+        let rep = request_build.send().await.unwrap();
+        let str = rep.text().await.unwrap();
+        let json = json!(str);
+        let access_token = json["access_token"].as_str().unwrap();
+        let refresh_token = json["refresh_token"].as_str().unwrap();
+        let expires_in = json["expires_in"].as_i64().unwrap();
+
+        user.access_token = access_token.to_string();
```

Comment:
Instead of modifying the user, try to create a new user here. You can use shorthand update syntax here:

```rust
fn refresh_token(&self, user: &User) -> Result<User, Error> {
   // ...
   Ok(User {
      access_token: access_token.to_string();
      refresh_token: refresh_token.to_string();
      ..user
   })
}
```

https://doc.rust-lang.org/reference/expressions/struct-expr.html#functional-update-syntax